src/messaging/ExchangeContext.cpp

```diff
@@ -238,6 +219,22 @@ ExchangeContext::ExchangeContext(ExchangeManager * em, uint16_t ExchangeId, Secu
     mFlags.Set(Flags::kFlagInitiator, Initiator);
     mDelegate = delegate;
 
+    ExchangeMessageDispatch * dispatch = nullptr;
+    if (delegate != nullptr)
+    {
+        // If there's a delegate, get the dispatch object from it.
+        dispatch = delegate->GetMessageDispatch(em->GetReliableMessageMgr(), em->GetSessionMgr());
+    }
+
+    // If the delegate is not provided, or the delegate hasn't provided a dispatch object, let's use
+    // the default delegate object
+    if (dispatch == nullptr)
+    {
+        dispatch = &em->mDefaultExchangeDispatch;
+        VerifyOrDie(dispatch != nullptr);
```

Comment:
VerifyOrDie should probably be outside if/else so that it applies to GetmessageDispatch as well (otherwise the retain would segfault anyway)

I am concerned that we do work in the constructor now: generally I expect constructors to only initialize variables. The pattern of having logic in constructors forces us to do 'Die' instead of being able to return an error code. This makes things a lot harder to debug (app restarts on embedded instead of printing out some sort of error log)